internal/pkg/vm/actor/builtin/power/power.go

```diff
@@ -0,0 +1,308 @@
+			// Find entry to delete.
+			var delEntry TableEntry
+			err := lookup.Find(ctx, delAddr.String(), &delEntry)
+			if err != nil {
+				if err == hamt.ErrNotFound {
+					return Errors[ErrUnknownEntry]
+				}
+				return errors.FaultErrorWrapf(err, "Could not retrieve power table entry with ID: %s", delAddr.String())
+			}
+
+			// Never delete an entry that still has power
+			if delEntry.ActivePower.IsPositive() || delEntry.InactivePower.IsPositive() {
+				return Errors[ErrDeleteMinerWithPower]
+			}
+
+			// All clear to delete
+			return lookup.Delete(ctx, delAddr.String())
+		})
+		if err != nil {
+			return nil, err
+		}
+		state.PowerTable = newPowerTable
+		return nil, nil
+	})
+	if err != nil {
+		return errors.CodeError(err), err
+	}
+	return 0, nil
+}
+
+// GetTotalPower returns the total power (in bytes) held by all miners registered in the system
+func (*impl) getTotalPower(vmctx runtime.Runtime) (*types.BytesAmount, uint8, error) {
+	if err := vmctx.Charge(actor.DefaultGasCost); err != nil {
+		return nil, internal.ErrInsufficientGas, errors.RevertErrorWrap(err, "Insufficient gas")
+	}
+
+	var state State
+	ret, err := actor.WithState(vmctx, &state, func() (interface{}, error) {
+		ctx := context.Background()
+		total := types.NewBytesAmount(0)
+		err := actor.WithLookupForReading(ctx, vmctx.Storage(), state.PowerTable, func(lookup storage.Lookup) error {
+			return lookup.ForEachValue(ctx, TableEntry{}, func(k string, value interface{}) error {
+				entry, ok := value.(TableEntry)
+				if !ok {
+					return errors.NewFaultError("Expected TableEntry from power table lookup")
+				}
+				total = total.Add(entry.ActivePower)
+				total = total.Add(entry.InactivePower)
+				return nil
+			})
+		})
+		return total, err
+	})
+	if err != nil {
+		return nil, errors.CodeError(err), err
+	}
+	return ret.(*types.BytesAmount), 0, nil
+}
+
+// ProcessPowerReport updates a registered miner's power table entry according to the power report.
+func (*impl) processPowerReport(vmctx runtime.Runtime, report types.PowerReport, updateAddr address.Address) (uint8, error) {
+	if err := vmctx.Charge(actor.DefaultGasCost); err != nil {
+		return internal.ErrInsufficientGas, errors.RevertErrorWrap(err, "Insufficient gas")
+	}
+
+	var state State
+	_, err := actor.WithState(vmctx, &state, func() (interface{}, error) {
+		ctx := context.Background()
+		newPowerTable, err := actor.WithLookup(ctx, vmctx.Storage(), state.PowerTable, func(lookup storage.Lookup) error {
+			// Find entry to update.
+			var updateEntry TableEntry
+			err := lookup.Find(ctx, updateAddr.String(), &updateEntry)
+			if err != nil {
+				if err == hamt.ErrNotFound {
+					return Errors[ErrUnknownEntry]
+				}
+				return errors.FaultErrorWrapf(err, "Could not retrieve power table entry with ID: %s", updateAddr.String())
+			}
+			// All clear to update
+			return lookup.Set(ctx, updateAddr.String(), report)
```

Comment:
Thanks this will be covered in the security issue/TODO.  

The thing about this one is that we really need to keep it open to some external address like the miner worker in order for us to mock out power to maintain our daemon tests before we have a fully integrated storage miner module.  And that check itself is totally insecure and needs to move to the miner actor address check (as you suggest) after integration.  For that reason I'd prefer keeping this call more blatantly insecure so that we can't forget about the problem.